### PR TITLE
LLM updates to main

### PR DESCRIPTION
**Pull request by ZiyueXu77**
Fixes # .

### Description

cherry pick PRs https://github.com/NVIDIA/NVFlare/pull/4128 and https://github.com/NVIDIA/NVFlare/pull/4137 from 2.7 to main

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Quick tests passed locally by running `./runtest.sh`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated.

### CODE REVIEW

File: examples/advanced/llm_hf/client.py [hunk 1]
```diff
@@ -180,15 +183,15 @@ def main():
     if rank == 0:
         print(f"Dataset size: training {len(dataset_train)}, validation {len(dataset_valid)}")
     # record every 5% of the dataset
-    batch_size = 4
-    gra_accu_steps = 10
+    # Adjust batch size based on training mode
+    batch_size = 2 if args.train_mode.lower() == "sft" else 4
+    gra_accu_steps = 20 if args.train_mode.lower() == "sft" else 10
     logging_steps = int(len(dataset_train) / (20 * batch_size * gra_accu_steps))
```
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply examples/advanced/llm_hf/client.py]: skip this for now, we may do a one-time update to all dataloader checks later

File: examples/advanced/llm_hf/utils/hf_sft_peft_iter.py [hunk 2]
```diff
@@ -164,9 +174,10 @@ def main():
     # weights for each round - to show the weights are loaded correctly
     initial_model_path = os.path.join(args.output_path, "pytorch_model_initial.pth")
     if train_mode:
-        params = get_peft_model_state_dict(model)
+        # Save PEFT part only
+        params = get_peft_model_state_dict(trainer.model)
     else:
-        params = model.state_dict()
+        params = trainer.model.state_dict()
```
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply examples/advanced/llm_hf/utils/hf_sft_peft_iter.py:190]: good point! added PEFT check

File: examples/advanced/llm_hf/client.py [hunk 3]
```diff
@@ -219,9 +222,23 @@ def main():
...
+    else:
+        peft_config = None
     model.config.pretraining_tp = 1
 
+    # Calculate warmup_steps (replacing deprecated warmup_ratio for future compatibility)
+    # Total training steps = (dataset_size / (batch_size * grad_accum * world_size)) * num_epochs
+    total_train_steps = (len(dataset_train) // (batch_size * gra_accu_steps * world_size)) * (
+        args.local_epoch * args.num_rounds
+    )
+    warmup_steps = int(total_train_steps * 0.03)  # 3% warmup
```
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply examples/advanced/llm_hf/client.py:237]: makes sense, updated to ensure at least 1 step

**Comment by ZiyueXu77** [comment]
@greptileai, auto suggestion commit is generating too much trouble, do not enable this function unless it is 100% correct

File: examples/advanced/llm_hf/README.md [hunk 4]
```diff
@@ -320,7 +320,7 @@ The SFT curves are shown below, magenta for centralized results, others for FL t
 
 These results show that model precision conversion / quantization does not significantly impact the training while reducing the message size to 1/2, 1/4, and even 1/8, which can significantly reduce the message size, making it crucial for transmitting LLM updates.
 
-For message reduce, from float32 to 16-/8-/4-bit, the message size (in MB) of GPT-Neo-1.3B model are reduced to: 
+For message reduce, from float32 to 16-/8-/4-bit, the message sizes (in MB) of GPT-Neo-1.3B model under SFT are reduced to:
```
Comment (holgerroth) [inline examples/advanced/llm_hf/README.md:323]: Check the sentence here. It should also include the "(from client to server after one local round)" comment.

Comment (ZiyueXu77) [reply examples/advanced/llm_hf/README.md:323]: good point added below

File: examples/advanced/llm_hf/utils/hf_sft_peft.py [hunk 5]
```diff
@@ -113,16 +115,27 @@ def main():
...
     model.config.pretraining_tp = 1
 
+    # Calculate warmup_steps (replacing deprecated warmup_ratio for future compatibility)
+    # Use ceiling division to ensure at least 1 step per epoch for small datasets
+    steps_per_epoch = math.ceil(len(dataset_train) / (batch_size * gra_accu_steps))
+    total_train_steps = steps_per_epoch * 3
+    warmup_steps = int(total_train_steps * 0.03)  # 3% warmup
+
+    # Set TensorBoard logging directory via environment variable
+    os.environ["TENSORBOARD_LOGGING_DIR"] = os.path.join(args.output_path, "logs")
```
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply examples/advanced/llm_hf/utils/hf_sft_peft.py:129]: it is ok for Warmup step to be 0 for small dataset
